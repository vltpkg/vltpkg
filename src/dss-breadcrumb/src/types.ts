/**
 * The specificity of a breadcrumb, used for sorting.
 */
export type BreadcrumbSpecificity = {
  idCounter: number
  commonCounter: number
}

/**
 * Options for the higher-level comparing breadcrumbs function.
 */
export type InternalModifierComparatorOptions = {
  range?: string
}

/**
 * Options for comparing breadcrumbs.
 */
export type ModifierComparatorOptions = {
  semver?: string
}

/**
 * A valid item of a given breadcrumb.
 */
export type ModifierBreadcrumbItem = {
  comparator: (opts: ModifierComparatorOptions) => boolean
  name?: string
  value: string
  type: string
  importer: boolean
  prev: ModifierBreadcrumbItem | undefined
  next: ModifierBreadcrumbItem | undefined
}

/**
 * A breadcrumb is a linked list of items, where
 * each item has a value and a type.
 */
export interface ModifierBreadcrumb
  extends Iterable<ModifierBreadcrumbItem> {
  clear(): void
  comment: string | undefined
  first: ModifierBreadcrumbItem
  last: ModifierBreadcrumbItem
  single: boolean
  specificity: BreadcrumbSpecificity
  interactive: () => ModifierInteractiveBreadcrumb
}

/**
 * An interactive breadcrumb that holds state on what is the current item.
 */
export type ModifierInteractiveBreadcrumb = {
  current: ModifierBreadcrumbItem | undefined
  done: boolean
  next: () => ModifierInteractiveBreadcrumb
}
