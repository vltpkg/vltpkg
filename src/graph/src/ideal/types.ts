import type { DepID } from '@vltpkg/dep-id'
import type { PackageInfoClient } from '@vltpkg/package-info'
import type {
  AddImportersDependenciesMap,
  Dependency,
  RemoveImportersDependenciesMap,
} from '../dependencies.ts'
import type { Graph } from '../graph.ts'

export type BuildIdealAddOptions = {
  /**
   * A {@link AddImportersDependenciesMap} in which keys are {@link DepID}
   * linking to another `Map` in which keys are the dependency names and values
   * are {@link Dependency}. This structure represents dependencies that need
   * to be added to the importer represented by {@link DepID}.
   */
  add: AddImportersDependenciesMap
}

export type BuildIdealRemoveOptions = {
  /**
   * A {@link RemoveImportersDependenciesMap} object representing nodes to be
   * removed from the ideal graph. Each {@link DepID} key represents an
   * importer node and the `Set` of dependency names to be removed from its
   * dependency list.
   */
  remove: RemoveImportersDependenciesMap
}

export type BuildIdealFromGraphOptions = {
  /**
   * An initial {@link Graph} to start building from, adding nodes to any
   * missing edges and appending any new specs defined in `addSpecs`.
   */
  graph: Graph
}

export type BuildIdealPackageInfoOptions = {
  /**
   * A {@link PackageInfoClient} instance to read manifest info from.
   */
  packageInfo: PackageInfoClient
}
