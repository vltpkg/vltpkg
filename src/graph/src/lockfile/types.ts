import type { DepID } from '@vltpkg/dep-id'
import type { Spec, SpecOptions } from '@vltpkg/spec'
import type { Integrity, Manifest } from '@vltpkg/types'
import type { DependencyTypeShort } from '../dependencies.js'

/**
 * This is the main type definition for the contents of the
 * `vlt-lock.json` file.
 *
 * The nodes and edges information from the lockfile are used to reconstruct
 * a {@link Graph} representing a previous install.
 */
export type LockfileData = {
  options: SpecOptions
  nodes: Record<DepID, LockfileNode>
  edges: LockfileEdges
}

export const getFlagNumFromNode = (node: {
  optional?: boolean
  dev?: boolean
}) =>
  node.optional && node.dev ? LockfileNodeFlagDevOptional
  : node.optional ? LockfileNodeFlagOptional
  : node.dev ? LockfileNodeFlagDev
  : LockfileNodeFlagNone

export const getBooleanFlagsFromNum = (flags: LockfileNodeFlags) => ({
  dev: !!(flags & LockfileNodeFlagDev),
  optional: !!(flags & LockfileNodeFlagOptional),
})

export const LockfileNodeFlagNone = 0
export const LockfileNodeFlagOptional = 1
export const LockfileNodeFlagDev = 2
export const LockfileNodeFlagDevOptional = 3

/**
 * Bit flags indicating whether a node is optional and/or dev.
 */
export type LockfileNodeFlags = 0 | 1 | 2 | 3

/**
 * Lockfile representation of a node from the install graph.
 */
export type LockfileNode = [
  flags: LockfileNodeFlags,
  name?: string | null,
  integrity?: Integrity | null,
  resolved?: string | null,
  location?: string | null,
  manifest?: Manifest | null,
]

/**
 * Lockfile edges are stored as a record object where the key
 * is `${from.id} ${spec.name}` and the value is
 * `${type} ${spec.bareSpec} ${to.id | 'MISSING'}`
 *
 * Storing them in a record like this means that we are guaranteed to
 * never end up with duplicates, and a standard `JSON.stringify()`
 * will nicely print them out one line per edge.
 */
export type LockfileEdges = {
  [key: LockfileEdgeKey]: LockfileEdgeValue
}

/** `${from} ${dep name}` */
export type LockfileEdgeKey = `${DepID} ${string}`

/** `${type} ${spec} ${to}` */
export type LockfileEdgeValue =
  `${DependencyTypeShort} ${Spec['bareSpec']} ${DepID | 'MISSING'}`
