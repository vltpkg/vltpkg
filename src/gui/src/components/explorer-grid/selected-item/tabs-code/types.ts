import type { FsItem } from '@/lib/fetch-fs'

export type FsItemWithNone = Omit<
  Pick<FsItem, 'type' | 'name' | 'size'>,
  'type'
> & {
  type: FsItem['type'] | 'none'
}

export interface ItemAlerts {
  title: string
  description: string
}

export interface Error {
  origin: string
  cause: string
}

export type View = 'code' | 'preview'

export type Language =
  | 'markdown'
  | 'typescript'
  | 'javascript'
  | 'jsx'
  | 'tsx'
  | 'json'
  | 'css'
  | 'html'
  | 'text'

export type SortingDirection = 'asc' | 'desc' | 'unsorted'

export type FsItemSortKey = 'name' | 'type' | 'size' | null
