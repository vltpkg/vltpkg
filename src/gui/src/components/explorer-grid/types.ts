import type { QueryResponseEdge } from '@vltpkg/query'

/**
 * A looser representation of an edge, it contains optional properties that
 * are meant to be used alongside importer nodes where they might be missing.
 */
export type EdgeLoose = Pick<QueryResponseEdge, 'name'> & {
  type?: QueryResponseEdge['type']
  spec?: QueryResponseEdge['spec']
  from?: QueryResponseEdge['from']
  to?: QueryResponseEdge['to']
}

/**
 * Data to be presented in the UI in the form of Cards. This is used for
 * both the main result column, selected node and side panel views.
 */
export type GridItemData = EdgeLoose & {
  /**
   * The original dependency name as defined in its parent manifest.
   */
  depName?: string
  /**
   * An index value shared between installed and uninstalled dependencies
   * to keep track of the order they were added to the UI.
   */
  depIndex?: number
  /**
   * A unique identifier for the item. This will not translate to `node.to.id`
   * but rather concatenate names / ids to create a unique id per item.
   * Used as the unique `key` for React components.
   */
  id: string
  /**
   * Whether the current result interface is showing only results for a single
   * node. Used for dynamically building queries when clicking the UI.
   */
  sameItems?: boolean
  /**
   * How many resulting edges this item represents.
   */
  size: number
  /**
   * Whether this item is representing multiple edge results.
   */
  stacked: boolean
  /**
   * A title to be displayed in the Card UI.
   */
  title: string
  /**
   * A version to be used in the UI.
   */
  version: string
  /**
   * A list of labels to be displayed in the Card UI.
   */
  labels?: string[]
}

/**
 * Options used to create new Card grid items to be displayed in the UI.
 */
export type GridItemOptions = {
  item: GridItemData
  dependencies?: boolean
  highlight?: boolean
  selected?: boolean
  side?: boolean
}
