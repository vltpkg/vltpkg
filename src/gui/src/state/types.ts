import {
  type LockfileData,
  type GraphLike,
  type NodeLike,
  type EdgeLike,
} from '@vltpkg/graph'
import { type Query } from '@vltpkg/query'
import { type SpecOptionsFilled } from '@vltpkg/spec/browser'
import { type Integrity, type Manifest } from '@vltpkg/types'

export type Action = {
  updateActiveRoute: (route: State['activeRoute']) => void
  updatePreviousRoute: (route: State['activeRoute']) => void
  updateDashboard: (dashboard: State['dashboard']) => void
  updateGraph: (graph: State['graph']) => void
  updateQ: (q: State['q']) => void
  updateQuery: (query: State['query']) => void
  updateEdges: (edges: State['edges']) => void
  updateErrorCause: (errorCause: State['errorCause']) => void
  updateHasDashboard: (hasDashboard: State['hasDashboard']) => void
  updateLinePositionReference: (position: number) => void
  updateNodes: (nodes: State['nodes']) => void
  updateProjectInfo: (projectInfo: State['projectInfo']) => void
  updateSelectedNode: (node: State['selectedNode']) => void
  updateSpecOptions: (specOptions: State['specOptions']) => void
  updateStamp: () => void
  updateTheme: (theme: State['theme']) => void
  reset: () => void
  saveQuery: (item: SavedQuery) => void
  updateSavedQuery: (savedQuery: SavedQuery) => void
  deleteSavedQueries: (savedQueries: SavedQuery[]) => void
  saveQueryLabel: (queryLabel: QueryLabel) => void
  updateSavedQueryLabel: (queryLabel: QueryLabel) => void
  deleteSavedQueryLabels: (queryLabels: QueryLabel[]) => void
}

/**
 * Project information present in the transfer data object.
 */
export type ProjectInfo = {
  /**
   * Other tools used by this project, such as runtime and package mangager.
   */
  tools: DashboardTools[]
  /**
   * `true` if this package node_modules folder was installed
   * using the vlt client.
   */
  vltInstalled?: boolean
}

/**
 * Transfer data object used to send data from the cli.
 */
export type TransferData = {
  importers: RawNode[]
  lockfile: LockfileData
  projectInfo: ProjectInfo
}

export type RawNode = {
  importer: boolean
  id: string
  name: string
  version: string
  location: string
  manifest: Manifest
  projectRoot?: string
  integrity?: Integrity
  resolved?: string
  dev: boolean
  optional: boolean
}

/**
 * The main state object for the graph explorer.
 */
export type State = {
  /**
   * The current location.pathname (e.g. route) in the app.
   */
  activeRoute: string
  /**
   * The last route in the app.
   */
  previousRoute: string
  /**
   * List of projects to be displayed in the dashboard.
   */
  dashboard?: DashboardData
  /**
   * Current graph to be explored.
   */
  graph?: GraphLike
  /**
   * List of selected edges returned after querying the graph.
   */
  edges: EdgeLike[]
  /**
   * An informative message to be displayed when an error occurs.
   */
  errorCause: string
  /**
   * Whether the dashboard is enabled or not.
   */
  hasDashboard: boolean
  /**
   * A reference value to properly draw connections between nodes in the graph.
   */
  linePositionReference: number
  /**
   * List of selected nodes returned after querying the graph.
   */
  nodes: NodeLike[]
  /**
   * Information about the current project being explored.
   */
  projectInfo: ProjectInfo
  /**
   * Holds the reference to the {@link Query} instance object.
   */
  q?: Query
  /**
   * The query string typed by the user in the interface.
   */
  query: string
  // TODO: remove selectedNode as it's unused
  /**
   * Reference to a currently selected node.
   */
  selectedNode?: NodeLike
  /**
   * Spec options used for the current graph.
   */
  specOptions?: SpecOptionsFilled
  /**
   * A random string to control when graph data should be reloaded.
   */
  stamp: string
  /**
   * Store the current theme value.
   */
  theme: 'light' | 'dark'
  /**
   * Saved queries in localStorage.
   */
  savedQueries?: SavedQuery[]
  /**
   * Saved labels used for query tags in localStorage.
   */
  savedQueryLabels?: QueryLabel[]
}

export type DashboardTools =
  | 'vlt'
  | 'node'
  | 'deno'
  | 'bun'
  | 'npm'
  | 'pnpm'
  | 'yarn'
  | 'js'

export type DashboardLocation = {
  /**
   * The file system path to the directory.
   */
  path: string
  /**
   * A shortened, human readable path to be displayed in the dashboard.
   */
  readablePath: string
}

export type DashboardData = {
  /**
   * The reference current working directory.
   */
  cwd: string
  /**
   * The app version.
   */
  buildVersion: string
  /**
   * The default author name to be used when creating new projects.
   */
  defaultAuthor: string
  /**
   * A list of projects to be displayed in the dashboard.
   */
  projects: DashboardDataProject[]
  /**
   * A list of directories that are read to find projects
   * to display in the dashboard. New projects can also be created
   * in these folders.
   */
  dashboardProjectLocations: DashboardLocation[]
}

export type QueryLabel = {
  id: string
  color: Color
  name: string
  description: string
}

export type SavedQuery = {
  id: string
  name: string
  context: string
  query: string
  dateCreated: string
  dateModified: string
  labels?: QueryLabel[]
}

export type DashboardDataProject = {
  name: string
  readablePath: string
  path: string
  manifest: Manifest
  tools: DashboardTools[]
  mtime?: number | null
}

export type Color = `#${string}`
