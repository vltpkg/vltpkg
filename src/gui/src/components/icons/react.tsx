export const React = (props: React.SVGProps<SVGSVGElement>) => {
  return (
    <svg
      width="500"
      height="500"
      viewBox="0 0 500 500"
      fill="none"
      xmlns="http://www.w3.org/2000/svg"
      {...props}>
      <g clip-path="url(#clip0_2967_809)">
        <path
          d="m250.504 284.862c270.046 284.862 285.889 269.029 285.889 249.499c285.889 229.968 270.046 214.136 250.504 214.136c230.961 214.136 215.119 229.968 215.119 249.499c215.119 269.029 230.961 284.862 250.504 284.862z"
          fill="currentColor"
        />
        <path
          d="m250.501 321.96c355.363 321.96 440.371 289.522 440.371 249.509c440.371 209.495 355.363 177.058 250.501 177.058c145.639 177.058 60.6318 209.495 60.6318 249.509c60.6318 289.522 145.639 321.96 250.501 321.96z"
          stroke="currentColor"
          stroke-width="18.913"
        />
        <path
          d="m187.724 285.725c240.155 376.482 310.768 433.837 345.442 413.83c380.116 393.823 365.721 304.031 313.29 213.274c260.859 122.517 190.247 65.1621 155.572 85.1689c120.898 105.176 135.293 194.968 187.724 285.725z"
          stroke="currentColor"
          stroke-width="18.913"
        />
        <path
          d="m187.726 213.277c135.295 304.034 120.9 393.826 155.574 413.833c190.248 433.84 260.861 376.485 313.292 285.728c365.723 194.971 380.118 105.179 345.443 85.172c310.769 65.1652 240.157 122.52 187.726 213.277z"
          stroke="currentColor"
          stroke-width="18.913"
        />
      </g>
      <defs>
        <clipPath id="clip0_2967_809">
          <rect
            width="397"
            height="353"
            fill="currentColor"
            transform="translate(52 73)"
          />
        </clipPath>
      </defs>
    </svg>
  )
}
