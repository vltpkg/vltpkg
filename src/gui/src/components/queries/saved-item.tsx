import { useEffect, useState } from 'react'
import {
  type SavedQuery,
  type Action,
  type QueryLabel,
} from '@/state/types.js'
import { useGraphStore } from '@/state/index.js'
import { Input } from '@/components/ui/input.jsx'
import { Button } from '@/components/ui/button.jsx'
import { useToast } from '@/components/hooks/use-toast.js'
import { Checkbox } from '@/components/ui/checkbox.jsx'
import { Label } from '@/components/ui/label.jsx'
import { LabelBadge } from '@/components/labels/label-badge.jsx'
import { ArrowRight, ChevronsUpDown } from 'lucide-react'
import { LabelSelect } from '@/components/labels/label-select.jsx'
import {
  Popover,
  PopoverTrigger,
  PopoverContent,
} from '@/components/ui/popover.jsx'
import {
  Tooltip,
  TooltipProvider,
  TooltipTrigger,
  TooltipContent,
} from '@/components/ui/tooltip.jsx'

type SelectQueryOptions = {
  updateActiveRoute: Action['updateActiveRoute']
  updateErrorCause: Action['updateErrorCause']
  updateQuery: Action['updateQuery']
  updateStamp: Action['updateStamp']
  item: SavedQuery
  context: string
}

// TODO: should reuse / share the project select logic from:
// src/gui/src/components/dashboard-grid/index.tsx
export const selectQuery = async ({
  updateActiveRoute,
  updateErrorCause,
  updateQuery,
  updateStamp,
  item,
  context,
}: SelectQueryOptions) => {
  let req
  try {
    req = await fetch('/select-project', {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify({
        path: context,
      }),
    })
  } catch (err) {
    console.error(err)
    updateActiveRoute('/error')
    updateErrorCause('Failed to request project selection.')
    return
  }

  let projectSelected = false
  try {
    projectSelected = (await req.json()) === 'ok'
  } catch (err) {
    console.error(err)
  }

  if (projectSelected) {
    window.scrollTo(0, 0)
    updateQuery(item.query)
    updateActiveRoute('/explore')
    updateStamp()
  } else {
    updateActiveRoute('/error')
    updateErrorCause('Failed to select project.')
  }
}

const SavedQueryItem = ({
  item,
  handleSelect,
  checked,
}: {
  item: SavedQuery
  handleSelect: (item: SavedQuery) => void
  checked: boolean
}) => {
  const { toast } = useToast()
  const [isExpanded, setIsExpanded] = useState<boolean>(false)
  const [editName, setEditName] = useState<string>('')
  const [editContext, setEditContext] = useState<string>('')
  const [_labelSelectOpen, setLabelSelectOpen] =
    useState<boolean>(false)
  const [popoverOpen, setPopoverOpen] = useState<boolean>(false)
  const [selectedLabels, setSelectedLabels] = useState<QueryLabel[]>(
    [],
  )
  const [editQuery, setEditQuery] = useState<string>('')
  const [isValid, setIsValid] = useState<boolean>(false)
  const updateSavedQuery = useGraphStore(
    state => state.updateSavedQuery,
  )
  const updateActiveRoute = useGraphStore(
    state => state.updateActiveRoute,
  )
  const updateErrorCause = useGraphStore(
    state => state.updateErrorCause,
  )
  const updateQuery = useGraphStore(state => state.updateQuery)
  const updateStamp = useGraphStore(state => state.updateStamp)

  useEffect(() => {
    setEditName(item.name)
    setEditContext(item.context)
    setEditQuery(item.query)
    setSelectedLabels(item.labels ?? [])
  }, [item])

  useEffect(() => {
    if (editName !== '' && editQuery !== '') {
      setIsValid(true)
    } else {
      setIsValid(false)
    }
  }, [editName, editContext, editQuery])

  const handleEdit = () => {
    setIsExpanded(!isExpanded)
  }

  const handleSaveChanges = () => {
    const updatedItem: SavedQuery = {
      ...item,
      name: editName,
      context: editContext,
      query: editQuery,
      dateModified: new Date().toISOString(),
      labels: selectedLabels,
    }
    updateSavedQuery(updatedItem)
    setIsExpanded(false)
    toast({
      title: `${editName} saved successfully`,
    })
  }

  const runQuery = async (): Promise<void> => {
    await selectQuery({
      updateActiveRoute,
      updateErrorCause,
      updateQuery,
      updateStamp,
      item,
      context: item.context,
    })
  }

  return (
    <div
      className={`group flex flex-col rounded-sm border border-[1px] bg-neutral-50 transition-all hover:border-foreground/50 dark:bg-neutral-950 ${isExpanded ? 'border-foreground/50' : 'border-muted-foreground/25'}`}>
      <div className="grid grid-cols-12 gap-4 px-3 py-2">
        <div className="col-span-2 flex grow items-center gap-3">
          <Checkbox
            checked={checked}
            onCheckedChange={() => handleSelect(item)}
            className="border-muted-foreground/25 group-hover:border-muted-foreground/50"
          />
          <p className="truncate text-sm font-medium">
            {editName.trim() !== '' ? editName : 'Query Name'}
          </p>
        </div>
        <div className="col-span-4 flex items-center">
          {!isExpanded && (
            <p className="w-[500px] truncate text-sm">{editQuery}</p>
          )}
        </div>
        <div className="col-span-2 flex w-full items-center gap-2">
          <TooltipProvider>
            <Tooltip>
              <TooltipTrigger className="w-full truncate text-left text-sm">
                {item.context}
              </TooltipTrigger>
              <TooltipContent>{item.context}</TooltipContent>
            </Tooltip>
          </TooltipProvider>
        </div>
        <div className="col-span-2 flex items-center gap-2">
          {selectedLabels.map((label, idx) => (
            <LabelBadge
              key={idx}
              name={label.name}
              color={label.color}
            />
          ))}
        </div>

        <div className="col-span-2 flex items-center justify-end gap-4">
          <Button
            variant="outline"
            className="h-[2rem] rounded-sm border border-[1px] border-muted-foreground/25 px-3 text-sm"
            onClick={handleEdit}>
            {isExpanded ? 'Close' : 'Edit'}
          </Button>
          {editContext.trim() !== '' && (
            <Button
              variant="outline"
              className="h-[2rem] rounded-sm border border-[1px] border-muted-foreground/25 px-3 text-sm"
              role="link"
              onClick={() => void runQuery()}>
              <span>Run</span>
              <ArrowRight />
            </Button>
          )}
        </div>
      </div>

      {/* expanded */}
      {isExpanded && (
        <div className="flex flex-col border-t-[1px] border-muted-foreground/25 px-3 py-2">
          {/* form */}
          <div className="my-3 mb-6 flex gap-3">
            <div className="flex flex-col gap-2">
              <Label className="border-none text-sm font-medium">
                Name
              </Label>
              <Input
                type="text"
                value={editName}
                onChange={e => setEditName(e.target.value)}
                placeholder="Name"
              />
            </div>
            <div className="flex grow flex-col gap-2">
              <Label className="border-none text-sm font-medium">
                Query
              </Label>
              <Input
                type="text"
                placeholder="Query"
                value={editQuery}
                onChange={e => setEditQuery(e.target.value)}
                className="w-full"
              />
            </div>
            <div className="flex grow flex-col gap-2">
              <Label className="border-none text-sm font-medium">
                Directory
              </Label>
              <Input
                type="text"
                value={editContext}
                onChange={e => setEditContext(e.target.value)}
                placeholder="Directory (optional)"
              />
            </div>
            <div className="flex grow flex-col gap-2">
              <Label className="border-none text-sm">Labels</Label>
              <Popover
                open={popoverOpen}
                onOpenChange={setPopoverOpen}>
                <PopoverTrigger asChild>
                  <Button
                    variant="outline"
                    role="combobox"
                    aria-expanded={popoverOpen}
                    className="w-full justify-between">
                    Select labels
                    <ChevronsUpDown className="opacity-50" />
                  </Button>
                </PopoverTrigger>
                <PopoverContent className="p-0">
                  <LabelSelect
                    selectedItems={selectedLabels}
                    setItems={setSelectedLabels}
                    setIsOpen={setLabelSelectOpen}
                  />
                </PopoverContent>
              </Popover>
            </div>
          </div>

          {/* footer */}
          <div className="mb-0.5 flex items-end justify-end">
            <Button
              disabled={!isValid}
              onClick={handleSaveChanges}
              variant="default"
              size="sm">
              Save changes
            </Button>
          </div>
        </div>
      )}
    </div>
  )
}

export { SavedQueryItem }
