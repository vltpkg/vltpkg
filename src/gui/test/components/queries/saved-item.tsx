import { vi, expect, afterEach, describe, it } from 'vitest'
import { cleanup, render } from '@testing-library/react'
import html from 'diffable-html'
import { useGraphStore as useStore } from '@/state/index.js'
import { type SavedQuery } from '@/state/types.js'
import { SavedQueryItem } from '@/components/queries/saved-item.jsx'

vi.mock('@/components/ui/input.jsx', () => ({
  Input: 'gui-label',
}))

vi.mock('@/components/ui/button.jsx', () => ({
  Button: 'gui-button',
}))

vi.mock('@/components/ui/checkbox.jsx', () => ({
  Checkbox: 'gui-checkbox',
}))

vi.mock('@/components/ui/label.jsx', () => ({
  Label: 'gui-label',
}))

vi.mock('@/components/labels/label-badge.jsx', () => ({
  LabelBadge: 'gui-label-badge',
}))

vi.mock('lucide-react', () => ({
  ArrowRight: 'gui-arrow-right-icon',
  ChevronsUpDown: 'gui-chevrons-up-down-icon',
}))

vi.mock('@/components/labels/label-select.jsx', () => ({
  LabelSelect: 'gui-label-select',
}))

vi.mock('@/components/ui/popover.jsx', () => ({
  Popover: 'gui-popover',
  PopoverTrigger: 'gui-popover-trigger',
  PopoverContent: 'gui-popover-content',
}))

vi.mock('@/components/ui/tooltip.jsx', () => ({
  Tooltip: 'gui-tooltip',
  TooltipProvider: 'gui-tooltip-provider',
  TooltipTrigger: 'gui-tooltip-trigger',
  TooltipContent: 'gui-tooltip-content',
}))

expect.addSnapshotSerializer({
  serialize: v => html(v),
  test: () => true,
})

afterEach(() => {
  const CleanUp = () => (useStore(state => state.reset)(), '')
  render(<CleanUp />)
  cleanup()
})

describe('labels view', () => {
  it('labels render default', () => {
    const mockHandleSelect = vi.fn()

    const mockSavedQuery: SavedQuery = {
      id: '98074a4c-6c33-4880-9412-9e0eaf8db609',
      name: 'mock-saved-query',
      context: '/mock/context/query',
      query: 'mock-query',
      dateCreated: 'date-created',
      dateModified: 'date-modified',
      labels: [],
    }

    const Container = () => {
      return (
        <SavedQueryItem
          item={mockSavedQuery}
          handleSelect={mockHandleSelect}
          checked={false}
        />
      )
    }

    const { container } = render(<Container />)
    expect(container.innerHTML).toMatchSnapshot()
  })
})
