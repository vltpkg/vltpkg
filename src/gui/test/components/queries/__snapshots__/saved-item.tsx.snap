// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`labels view > labels render default 1`] = `

<div class="group flex flex-col rounded-sm border border-[1px] bg-neutral-50 transition-all hover:border-foreground/50 dark:bg-neutral-950 border-muted-foreground/25">
  <div class="grid grid-cols-12 gap-4 px-3 py-2">
    <div class="col-span-2 flex grow items-center gap-3">
      <gui-checkbox
        checked="false"
        classname="border-muted-foreground/25 group-hover:border-muted-foreground/50"
      >
      </gui-checkbox>
      <p class="truncate text-sm font-medium">
        mock-saved-query
      </p>
    </div>
    <div class="col-span-4 flex items-center">
      <p class="w-[500px] truncate text-sm">
        mock-query
      </p>
    </div>
    <div class="col-span-2 flex w-full items-center gap-2">
      <gui-tooltip-provider>
        <gui-tooltip>
          <gui-tooltip-trigger classname="w-full truncate text-left text-sm">
            /mock/context/query
          </gui-tooltip-trigger>
          <gui-tooltip-content>
            /mock/context/query
          </gui-tooltip-content>
        </gui-tooltip>
      </gui-tooltip-provider>
    </div>
    <div class="col-span-2 flex items-center gap-2">
    </div>
    <div class="col-span-2 flex items-center justify-end gap-4">
      <gui-button
        variant="outline"
        classname="h-[2rem] rounded-sm border border-[1px] border-muted-foreground/25 px-3 text-sm"
      >
        Edit
      </gui-button>
      <gui-button
        variant="outline"
        classname="h-[2rem] rounded-sm border border-[1px] border-muted-foreground/25 px-3 text-sm"
        role="link"
      >
        <span>
          Run
        </span>
        <gui-arrow-right-icon>
        </gui-arrow-right-icon>
      </gui-button>
    </div>
  </div>
</div>

`;
